File: app/src/main/java/com/cyl/musiclake/ui/music/list/activity/PlaylistDetailActivity.java
package com.cyl.musiclake.ui.music.list.activity;

import android.content.Intent;
import android.os.Build;
import android.support.annotation.RequiresApi;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.transition.Transition;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;

import com.cyl.musiclake.R;
import com.cyl.musiclake.base.BaseActivity;
import com.cyl.musiclake.bean.Music;
import com.cyl.musiclake.common.Extras;
import com.cyl.musiclake.data.PlaylistLoader;
import com.cyl.musiclake.ui.music.list.adapter.SongAdapter;
import com.cyl.musiclake.ui.music.search.SearchActivity;
import com.cyl.musiclake.ui.zone.EditActivity;
import com.cyl.musiclake.utils.CoverLoader;
import com.cyl.musiclake.utils.ImageUtils;
import com.cyl.musiclake.utils.SystemUtils;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;

/**
 * 作者：yonglong on 2016/8/14 16:15
 * 邮箱：643872807@qq.com
 * 版本：2.5
 */
public class PlaylistDetailActivity extends BaseActivity {

    @BindView(R.id.recyclerView)
    RecyclerView mRecyclerView;
    @BindView(R.id.album_art)
    ImageView album_art;
    @BindView(R.id.foreground)
    View foreground;

    private SongAdapter mAdapter;
    private List<Music> musicInfos = new ArrayList<>();
    private String pid;
    private long album_id;
    //0代表专辑，1代表艺术家
    private int isAlbum;

    /**
     * 设置监听事件
     */
    @Override
    protected void listener() {


    }

    @Override
    protected int getLayoutResID() {
        return R.layout.frag_playlist_detail;
    }

    private void setAlbumart() {
        setToolbarTitle(getIntent().getExtras().getString(Extras.PLAYLIST_NAME));
        if (isAlbum == -1) {
            Log.e("====", getIntent().getExtras().getString(Extras.PLAYLIST_FOREGROUND_COLOR) + "==\n"
            );
            foreground.setBackgroundColor(getIntent().getExtras().getInt(Extras.PLAYLIST_FOREGROUND_COLOR));
            album_art.setBackgroundResource(getIntent().getExtras().getInt(Extras.PLAYLIST_BACKGROUND_IMAGE));
        } else {
            loadBitmap(ImageUtils.getAlbumArtUri(album_id).toString());
        }
    }

    private void loadBitmap(String uri) {
        Log.e("EEEE", uri);
        CoverLoader.loadImageView(this, uri, album_art);
    }

    @Override
    protected void initData() {

        mRecyclerView.setLayoutManager(new LinearLayoutManager(this));
        pid = getIntent().getStringExtra(Extras.PLAYLIST_ID);
        isAlbum = getIntent().getIntExtra(Extras.ALBUM, -1);
        album_id = getIntent().getLongExtra(Extras.ALBUM_ID, -1);
        Log.e("pid", pid + "===" + isAlbum + "+++" + album_id + "00");
        setAlbumart();
    }

    @Override
    protected void initView() {
        if (SystemUtils.isLollipop()) {
            getWindow().getEnterTransition().addListener(new EnterTransition());
        }
    }


    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        switch (id) {
            case R.id.action_search:
                final Intent intent = new Intent(this, SearchActivity.class);
                intent.setFlags(Intent.FLAG_ACTIVITY_NO_ANIMATION);
                startActivity(intent);
                break;
            case R.id.action_delete_playlist:
                PlaylistLoader.deletePlaylist(this, pid);
                finish();
                break;
            case R.id.action_share:
                Intent intent3 = new Intent(PlaylistDetailActivity.this, EditActivity.class);
                String content = "";
                if (musicInfos.size() > 0) {
                    content = "分享歌单\n";
                }
                for (int i = 0; i < musicInfos.size(); i++) {
                    content += musicInfos.get(i).getTitle() + "---" + musicInfos.get(i).getArtist();
                    content += "\n";
                }
                intent3.putExtra("content", content);
                startActivity(intent3);
                break;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        if (isAlbum != -1) {
            getMenuInflater().inflate(R.menu.menu_playlist, menu);
        } else {
            getMenuInflater().inflate(R.menu.menu_playlist_detail, menu);
        }
        return super.onCreateOptionsMenu(menu);

    }


    @RequiresApi(api = Build.VERSION_CODES.KITKAT)
    private class EnterTransition implements Transition.TransitionListener {
        @Override
        public void onTransitionStart(Transition transition) {

        }

        @Override
        public void onTransitionEnd(Transition transition) {
        }

        @Override
        public void onTransitionCancel(Transition transition) {

        }

        @Override
        public void onTransitionPause(Transition transition) {

        }

        @Override
        public void onTransitionResume(Transition transition) {

        }
    }
}

File: app/src/main/java/com/cyl/musiclake/ui/music/playlist/detail/PlaylistDetailActivity.kt
package com.cyl.musiclake.ui.music.playlist.detail

import android.content.Intent
import android.text.InputType
import android.view.Menu
import android.view.MenuItem
import androidx.recyclerview.widget.LinearLayoutManager
import com.afollestad.materialdialogs.MaterialDialog
import com.afollestad.materialdialogs.input.getInputField
import com.afollestad.materialdialogs.input.input
import com.cyl.musiclake.R
import com.cyl.musiclake.bean.Album
import com.cyl.musiclake.bean.Artist
import com.cyl.musiclake.bean.Music
import com.cyl.musiclake.bean.Playlist
import com.cyl.musiclake.common.Constants
import com.cyl.musiclake.common.Extras
import com.cyl.musiclake.data.PlayHistoryLoader
import com.cyl.musiclake.event.PlaylistEvent
import com.cyl.musiclake.player.PlayManager
import com.cyl.musiclake.ui.base.BaseActivity
import com.cyl.musiclake.ui.deletePlaylist
import com.cyl.musiclake.ui.music.dialog.BottomDialogFragment
import com.cyl.musiclake.ui.music.edit.EditSongListActivity
import com.cyl.musiclake.ui.music.edit.PlaylistManagerUtils
import com.cyl.musiclake.ui.music.local.adapter.SongAdapter
import com.cyl.musiclake.ui.music.search.PlaylistSearchActivity
import com.cyl.musiclake.ui.my.BindLoginActivity
import com.cyl.musiclake.ui.widget.ItemDecoration
import com.cyl.musiclake.utils.CoverLoader
import com.cyl.musiclake.utils.LogUtil
import kotlinx.android.synthetic.main.frag_playlist_detail.*
import org.greenrobot.eventbus.EventBus
import org.jetbrains.anko.startActivity

/**
 * 作者：yonglong on 2016/8/15 19:54
 * 邮箱：643872807@qq.com
 * 版本：2.5
 * 歌单详情页
 */
class PlaylistDetailActivity : BaseActivity<PlaylistDetailPresenter>(), PlaylistDetailContract.View {
    /**
     * 显示异常UI
     */
    override fun showErrorTips(msg: String, hasTry: Boolean) {
        showError(msg, hasTry)
    }

    override fun showEmptyView(msg: String) {

    }

    override fun showTitle(title: String) {
        updateTitle(title)
    }

    override fun showCover(cover: String) {
        CoverLoader.loadImageView(context, coverUrl, album_art)
    }

    override fun showDescInfo(title: String) {

    }

    private var mAdapter: SongAdapter? = null
    private val musicList = mutableListOf<Music>()
    private var mPlaylist: Playlist? = null
    private var mArtist: Artist? = null
    private var pid: String? = null
    private var mAlbum: Album? = null
    private var title: String? = null
    private var coverUrl: String? = null
    private var type: String? = null

    private var bottomDialogFragment: BottomDialogFragment? = null

    override fun getLayoutResID(): Int {
        return R.layout.frag_playlist_detail
    }

    override fun initData() {
        showLoading()
        mPlaylist?.let {
            if (it.musicList.size > 0) {
                showPlaylistSongs(it.musicList)
            } else {
                mPresenter?.loadPlaylistSongs(it)
            }
        }
        mArtist?.let {
            mPresenter?.loadArtistSongs(it)
        }
        mAlbum?.let {
            mPresenter?.loadAlbumSongs(it)
        }
    }

    /**
     * 歌单页标题
     */
    override fun setToolbarTitle(): String? {
        mPlaylist = intent.getParcelableExtra(Extras.PLAYLIST)
        mArtist = intent.getParcelableExtra(Extras.ARTIST)
        mAlbum = intent.getSerializableExtra(Extras.ALBUM) as Album?
        mPlaylist?.let {
            title = it.name
            pid = it.pid
            coverUrl = it.coverUrl
            CoverLoader.loadBigImageView(context, coverUrl, it.type, album_art)
        }
        mArtist?.let {
            title = it.name
            pid = it.artistId.toString()
            coverUrl = it.picUrl
            CoverLoader.loadBigImageView(context, coverUrl, it.type, album_art)
        }
        mAlbum?.let {
            title = it.name
            pid = it.albumId.toString()
            coverUrl = it.cover
            CoverLoader.loadBigImageView(context, coverUrl, it.type, album_art)
        }
        return title
    }

    override fun initView() {
        mAdapter = SongAdapter(mutableListOf())
        recyclerView.layoutManager = LinearLayoutManager(this)
        recyclerView.addItemDecoration(ItemDecoration(this, ItemDecoration.VERTICAL_LIST))
        recyclerView.adapter = mAdapter
        fab.setOnClickListener { PlayManager.play(0, musicList, pid) }
    }

    override fun initInjector() {
        mActivityComponent.inject(this)
    }

    override fun listener() {
        mAdapter?.setOnItemClickListener { _, view, position ->
            if (view.id != R.id.iv_more) {
                when {
                    mPlaylist != null -> PlayManager.play(position, musicList, mPlaylist?.pid)
                    mArtist != null -> PlayManager.play(position, musicList, mArtist?.artistId.toString())
                    mAlbum != null -> PlayManager.play(position, musicList, mAlbum?.albumId.toString())
                }
//                mAdapter?.notifyDataSetChanged()
            }
        }
        mAdapter?.setOnItemChildClickListener { _, _, position ->
            bottomDialogFragment = BottomDialogFragment.newInstance(musicList[position], mPlaylist?.type).apply {
                mPlaylist?.pid?.let {
                    pid = it
                }
            }
            bottomDialogFragment?.removeSuccessListener = {
                removeMusic(position)
            }
            bottomDialogFragment?.show(this)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_delete_playlist -> {
                LogUtil.e("action_delete_playlist")
                mPlaylist?.let {
                    deletePlaylist(it, success = {
                        if (mPlaylist?.type == Constants.PLAYLIST_HISTORY_ID) {
                            musicList.clear()
                            PlayHistoryLoader.clearPlayHistory()
                            mAdapter?.notifyDataSetChanged()
                            showEmptyState()
                            EventBus.getDefault().post(PlaylistEvent(Constants.PLAYLIST_HISTORY_ID, it))
                        } else if (mPresenter != null) {
                            PlaylistManagerUtils.deletePlaylist(it) {
                                onBackPress()
                            }
                        }
                    })
                }
            }
            R.id.action_rename_playlist -> {
                LogUtil.e("action_rename_playlist")
                mPlaylist?.let { playlist ->
                    MaterialDialog(this).show {
                        title(R.string.playlist_rename)
                        positiveButton(R.string.sure)
                        negativeButton(R.string.cancel)
                        input(
                            hintRes = R.string.input_playlist, maxLength = 10, prefill = playlist.name,
                            inputType = InputType.TYPE_CLASS_TEXT
                        ) { dialog, input ->
                            LogUtil.e("=====", input.toString())
                        }
                        positiveButton {
                            val title = it.getInputField().text.toString()
                            if (title != mPlaylist?.name) {
                                mPresenter?.renamePlaylist(playlist, title)
                            }
                        }
                    }
                }
            }
            R.id.action_batch -> {
                EditSongListActivity.musicList = musicList
                startActivity<EditSongListActivity>()
            }
            R.id.action_order_title -> {
                musicList.sortBy { it.title }
                mAdapter?.notifyDataSetChanged()
            }
            R.id.action_order_album -> {
                musicList.sortBy { it.album }
                mAdapter?.notifyDataSetChanged()
            }
            R.id.action_order_artist -> {
                musicList.sortBy { it.artist }
                mAdapter?.notifyDataSetChanged()
            }
            R.id.action_search -> {
                //设置歌单内搜索的歌曲
                PlaylistSearchActivity.musicList = musicList;
                val intent = Intent(this, PlaylistSearchActivity::class.java)
                intent.flags = Intent.FLAG_ACTIVITY_NO_ANIMATION
                startActivity(intent)
            }
        }
        return super.onOptionsItemSelected(item)

    }

    private fun onBackPress() {
        this.onBackPressed()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_playlist_detail, menu)
        if (mPlaylist == null) {
            menu.removeItem(R.id.action_rename_playlist)
            menu.removeItem(R.id.action_delete_playlist)
        }
        mPlaylist?.let {
            when (it.pid) {
                Constants.PLAYLIST_HISTORY_ID -> {
                    menu.removeItem(R.id.action_rename_playlist)
                }
                Constants.PLAYLIST_LOVE_ID -> {
                    menu.removeItem(R.id.action_rename_playlist)
                    menu.removeItem(R.id.action_delete_playlist)
                }
            }
            if (it.type == Constants.PLAYLIST_BD_ID) {
                //百度电台
                menu.removeItem(R.id.action_rename_playlist)
                menu.removeItem(R.id.action_delete_playlist)
            }
        }
        return super.onCreateOptionsMenu(menu)
    }


    override fun showPlaylistSongs(songList: MutableList<Music>?) {
        hideLoading()
        LogUtil.e("showPlaylistSongs size =${songList?.size}")
        songList?.let {
            musicList.addAll(songList)
        }
        mAdapter?.setNewInstance(musicList)
        if (coverUrl == null) {
            mPlaylist?.coverUrl?.let {
                coverUrl = it
            }
            mArtist?.picUrl?.let {
                coverUrl = it
            }
            if (musicList.size > 0 && coverUrl == null) {
                coverUrl = musicList[0].coverUri
                type = musicList[0].type
            }
            CoverLoader.loadBigImageView(context, coverUrl, type, album_art)
        }
        if (musicList.size == 0) {
            showEmptyState()
        }
    }

    override fun removeMusic(position: Int) {
        musicList.removeAt(position)
        mAdapter?.notifyItemRemoved(position)
        if (musicList.size == 0) {
            showEmptyState()
        }
    }

    override fun success(type: Int) {
        onBackPress()
    }

    /**
     * 点击重试按钮响应事件
     */
    override fun retryLoading() {
        super.retryLoading()
        if (mPlaylist?.type == Constants.PLAYLIST_WY_RECOMMEND_ID) {
            val intent = Intent(this, BindLoginActivity::class.java)
            startActivityForResult(intent, Constants.REQUEST_CODE_LOGIN)
            return
        }
        showLoading()
        mPlaylist?.let {
            if (it.musicList.size > 0) {
                showPlaylistSongs(it.musicList)
            } else {
                mPresenter?.loadPlaylistSongs(it)
            }
        }
        mArtist?.let {
            mPresenter?.loadArtistSongs(it)
        }
        mAlbum?.let {
            mPresenter?.loadAlbumSongs(it)
        }
    }
}
